### DISCUSSION TITLE
How does vs_library fix the issue with events at the same tick or something? And how would it be done without vs_library?

### DISCUSSION THREAD

**Discussion by karl-police** [samisalreadytaken/vs_library#7] [Q&A] [answered]
I have a bunch of bots in a stack

![image](https://user-images.githubusercontent.com/12023782/185436504-0430c6ba-6793-4929-992a-cf018221d2f6.png)

and I have this setup in hammer

![image](https://user-images.githubusercontent.com/12023782/185436597-773b98fe-d324-4fd3-a486-e53681055599.png)

a "logic_eventlistener"

``::OnGameEvent_player_hurt(event_data)``
![image](https://user-images.githubusercontent.com/12023782/185453549-f3ab90eb-5d18-4e18-aa3b-3d80ada48947.png)

I have a file called ``eventlistener1.nut``, which contains:

```
::OnGameEvent_player_say

::OnGameEvent_player_hurt
```

&nbsp;

The second file is loaded by the "logic_script" called ``eventlistener_testing1.nut``:

```
function foreachPrint(data) {
	foreach (k,v in data) {
		printl(k + ": " + v)
	}
}


::OnGameEvent_player_say <- function(event) {
	printl("== Event: player_say ==")
	foreachPrint(event)
	printl("--------")
}.bindenv(this)


::OnGameEvent_player_hurt <- function(event) {
	printl("== Event: player_hurt ==")
	foreachPrint(event)
	printl("--------")
}.bindenv(this)
```

This second file successfully overrides the functions that I've defined in ``eventlistener1.nut``. The reason why I did it like that, is because the "logic_eventlistener" needs a script to be inside "Entity Scripts", otherwise it gives me an error, saying that "event_data" wouldn't exist.

&nbsp;

I am looking at "player_hurt". It gets called, whenever a player gets hurt including bots.

&nbsp;

&nbsp;

So what I did, is I gave myself an AWP and shot multiple bots at the same time, as example two.
Which printed out this:

```
== Event: player_hurt ==
armor: 98
health: 0
dmg_armor: 1
dmg_health: 149
userid: 265
weapon: awp
hitgroup: 1
attacker: 12
--------
== Event: player_hurt ==
armor: 98
health: 0
dmg_armor: 1
dmg_health: 149
userid: 265
weapon: awp
hitgroup: 1
attacker: 12
--------
```

&nbsp;

**So what's the issue?**

I shot multiple different bots, and it only printed out the same values. The ``userid`` is supposed to be different and not the same.

What I am asking is, why does this happen and how does "vs_library" deal with it? And how would I fix this, without the use of "vs_library" ?

This is only to understand vs_library better and how you would deal with this issue without "vs_library".

How would I have to modify ``eventlistener1_testing.nut`` or something else, so that the values don't "duplicate send" or something, without "vs_library" ?

**Comment by karl-police** [comment]
@samisalreadytaken would be cool if you could answer this question :)

**Comment by samisalreadytaken** [comment, marked as answer]
> the reason why I do it like that, is because the "logic_eventlistener" needs a script to be inside "Entity Scripts", otherwise it gives me an error, saying that "event_data" wouldn't exist.

This is not right, there must be another issue going on.

> What I am asking is, how does "vs_library" deal with it?

Hooking the _newslot metamethod of the event listener's script scope.

> And how would I fix this, without the use of "vs_library" ?
> ... and how you would deal with this issue without "vs_library".

You basically don't. While you could start with the very first step of creating metamethods - which you would either copy-paste or create a script file to be included and referenced for each of your event listeners which you would have to create in Hammer then compile the map on each change, you will have other issues after that, fixing that then deeper issues. At that point I would be just repeating the `src/vs_events.nut` file line by line.

The [`vs_events.nut`](https://raw.githubusercontent.com/samisalreadytaken/vs_library/master/vs_events.nut) library is the lightest, most optimal and most flexible fix for this. If you don't want to use any of the vs_library features, you would use this file alone and manage all of your game event callbacks safely in script. It only includes the bare minimum functionality.
